Make data eviction based on user interaction time when ITP is enabled

szewai opened the pull request:
#### 781b9f59f164fdfb902bdbbe5c67f6472a2b8a0f
<pre>
Make data eviction based on user interaction time when ITP is enabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=255272">https://bugs.webkit.org/show_bug.cgi?id=255272</a>

Reviewed by Youenn Fablet.

Revamp data eviction from exceeding total quota. The major changes:
1. NetworkStorageManager tracks total usage and performs eviction by itself instead of relying on NetworkQuotaManager;
this is because NetworkStorageManager has a list of origins that are being used (m_originStorageManagers) and it&apos;s more
convenient to do operation on OriginStorageManager from NetworkStorageManager.
2. NetworkStorageManager uses how much space is granted to all origins as estimated total usage. It only calculates the
real usage when estimated usage is bigger than quota, and then uses real usage to decide if eviction is needed.
3. When ITP is enabled, NetworkStorageManager decides the eviction order of origins based on interaction time from ITP;
when ITP is not enabled, NetworkStorageManager decides the order based on last modification time of storage files or
directories.
4. NetworkStorageManager will update last modification time of storage files or directories of an origin when it is
granted space.

API test: WKWebsiteDataStoreConfiguration.TotalQuotaRatioWithResourceLoadStatisticsEnabled

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.cpp:
(WebKit::ResourceLoadStatisticsDatabaseStore::allDomainsWithLastAccessedTime const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::registrableDomainsWithLastAccessedTime):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::registrableDomainsWithLastAccessedTime):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
(WebKit::NetworkSession::NetworkSession):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::spaceGrantedForOrigin):
(WebKit::NetworkStorageManager::schedulePerformEviction):
(WebKit::NetworkStorageManager::prepareForEviction):
(WebKit::NetworkStorageManager::lastModificationTimeForOrigin const):
(WebKit::NetworkStorageManager::donePrepareForEviction):
(WebKit::NetworkStorageManager::performEviction):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::updateLastModificationTimeForOrigin):
(WebKit::NetworkStorageManager::quotaManager): Deleted.
(WebKit::NetworkStorageManager::updateOriginModificationTime): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::usage):
(WebKit::OriginQuotaManager::grantWithCurrentQuota):
(WebKit::OriginQuotaManager::spaceGranted):
(WebKit::OriginQuotaManager::grantFastPath):
(WebKit::OriginQuotaManager::usageUpdated): Deleted.
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
(WebKit::OriginQuotaManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/262900@main">https://commits.webkit.org/262900@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9ccff0705765ebad640333ed0c15d90acfcf307b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2935 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/3000 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/3097 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4340 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/3347 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2902 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/3075 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/3038 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2571 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2963 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/3342 "2 failures") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2624 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/4132 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/838 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2605 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/2466 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2600 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2656 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3877 "261 api tests failed or timed out") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/3004 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2408 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2639 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2608 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/731 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2628 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2825 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.cpp:
```diff
@@ -2024,6 +2024,20 @@ Vector<RegistrableDomain> ResourceLoadStatisticsDatabaseStore::allDomains() cons
+    auto domainDatas = domains();
+    for (auto& domainData : domainDatas) {
+        auto lastAccessedTime = domainData.mostRecentUserInteractionTime > domainData.mostRecentWebPushInteractionTime ? domainData.mostRecentUserInteractionTime : domainData.mostRecentWebPushInteractionTime;
+        result.add(domainData.registrableDomain, lastAccessedTime);
```
result.add(domainData.registrableDomain, std::max(domainData.mostRecentUserInteractionTime, domainData.mostRecentWebPushInteractionTime ));

youennf commented on Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
```diff
@@ -664,18 +664,18 @@ Ref<OriginQuotaManager> OriginStorageManager::createQuotaManager()
         return IDBStorageManager::idbStorageSize(idbStoragePath) + CacheStorageManager::cacheStorageSize(cacheStoragePath) + fileSystemStorageSize;
     };
-    return OriginQuotaManager::create(m_quota, m_standardReportedQuota, WTFMove(getUsageFunction), std::exchange(m_increaseQuotaFunction, { }), std::exchange(m_notifyUsageUpdateFunction, { }));
+    return OriginQuotaManager::create(m_quota, m_standardReportedQuota, WTFMove(getUsageFunction), std::exchange(m_increaseQuotaFunction, { }), std::exchange(m_notifySpaceGrantedFunction, { }));
```
It is pre-existing but it is a bit strange to use temp OriginStorageManager variables.

szewai replied:
I don't quite get this comment; what do you mean by using temp OriginStorageManager variables?

youennf replied:
We are setting all these m_XXFunction(s) in OriginStorageManager, just to wait for creating OriginQuotaManager. This is a bit odd but ok.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -454,22 +567,23 @@ bool NetworkStorageManager::removeOriginStorageManagerIfPossible(const WebCore::
 {
     assertIsCurrent(workQueue());
-    // This function should be called when origin is in use, i.e. OriginStorageManager exists.
+    // This function must be called when origin is in use, i.e. OriginStorageManager exists.
```
Let's add ASSERT(manager) then

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
-                quota = *m_totalQuotaRatio * *volumeCapacity;
+        if (!canDelete) {
+            RELEASE_LOG(Storage, "%p - NetworkStorageManager::performEviction evicts %" PRIu64 " origins before getting an active origin", this, deletedOriginCount);
+            return;
```
Why stop here?
Also, didGetDomainsWithLastAccessedTimeForEviction creates originStorageManagers and calls removeOriginStorageManagerIfPossible, so we already know whether we would exit early.

szewai replied:
Stop here because I am not sure whether we should remove data of a "more-recently-used" origin when we fail to delete a not so recently used one.

ah it's a good point we should already know whether an origin is active in `didGetDomainsWithLastAccessedTimeForEviction`; will update.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        lastModificationTime = localStorageModificationTime > lastModificationTime ? localStorageModificationTime : lastModificationTime;
+        auto idbStoragePath = IDBStorageManager::idbStorageOriginDirectory(m_customIDBStoragePath, origin);
+        auto idbStorageModificationTime = valueOrDefault(FileSystem::fileModificationTime(idbStoragePath));
+        lastModificationTime = idbStorageModificationTime > lastModificationTime ? idbStorageModificationTime : lastModificationTime;
```
return std::max()

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    if (m_unifiedOriginStorageLevel <= UnifiedOriginStorageLevel::Basic) {
+        auto cacheStoragePath = CacheStorageManager::cacheStorageOriginDirectory(m_customCacheStoragePath, origin);
+        auto cacheStorageModificationTime = valueOrDefault(FileSystem::fileModificationTime(cacheStoragePath));
+        lastModificationTime = cacheStorageModificationTime > lastModificationTime ? cacheStorageModificationTime : lastModificationTime;
```
return std::max()

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    if (m_unifiedOriginStorageLevel <= UnifiedOriginStorageLevel::Standard) {
+        auto originFile = manager.path();
+        auto originFileModificationTime = valueOrDefault(FileSystem::fileModificationTime(originFile));
+        lastModificationTime = originFileModificationTime > lastModificationTime ? originFileModificationTime : lastModificationTime;
```
return std::max

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        lastModificationTime = originFileModificationTime > lastModificationTime ? originFileModificationTime : lastModificationTime;
+    }
+
+    return lastModificationTime;
```
ASSERT_NOT_REACHED();
return { };

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
 
-        std::sort(origins.begin(), origins.end(), [&](const auto& a, const auto& b) {
-            return a.second < b.second;
+        m_process->registrableDomainsWithLastAccessedTime(m_sessionID, [this, weakThis = ThreadSafeWeakPtr { *this }](auto result) mutable {
```
Why using weakThis here?
Should we just pass the ref along?
Maybe weakthis = WTFMove(weakThis).

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+            if (!strongThis || m_closed)
+                return;
+
+            m_queue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, result = crossThreadCopy(result)]() mutable {
```
Ditto.
Also crossThreadCopy(WTFMove(result))

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+                return;
+
+            m_queue->dispatch([weakThis = ThreadSafeWeakPtr { *this }, result = crossThreadCopy(result)]() mutable {
+                if (auto strongThis = weakThis.get())
```
We probably want to ensure NetworkStorageManager is destroyed in main thread.
We then need to move strongThis back to main thread.

szewai replied:
`NetworkStorageManager` will be destroyed in the main thread because `NetworkStorageManager::close()`, which sets `m_closed` should dispatch a last task (the last task dispatches a last task reply to main thread), and we have assert in `~NetworkStorageManager`.

Note this function does not dispatch task to m_queue when `m_closed` is true.

youennf replied:
Then, why should we even have a weakThis/strongThis approach?
It seems we could just capture 'this', maybe add a postTask so that we understand this is fine to do like in other parts of the code base.

szewai replied:
This is good point; I will file a bug to make other places to only capture `this`.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+        return;
+
+    m_isEvictionScheduled = true;
+    getDomainsWithLastAccessedTimeForEviction();
```
This is not really about getting domains but more about preparing eviction?
How about renaming to prepareDomainsEviction or something similar?

szewai replied:
Yes, that's why it has `ForEviction`; will rename it.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    return lastModificationTime;
+}
+
+void NetworkStorageManager::didGetDomainsWithLastAccessedTimeForEviction(std::optional<HashMap<WebCore::RegistrableDomain, WallTime>> domainsWithLastAccessedTime)
```
I would tend to rename it to performDomainsEviction.
And use a const std::optional&

szewai replied:
There is a `performEviction` below; I will rename it to `donePrepareForEviction` to match `prepareForEviction`.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
 #endif
+}
+
+void NetworkStorageManager::spaceGrantedForOrigin(const WebCore::ClientOrigin& origin, uint64_t spaceRequested)
```
s/spaceRequested/amount/

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+WallTime NetworkStorageManager::lastModificationTimeForOrigin(const WebCore::ClientOrigin& origin, OriginStorageManager& manager)
+{
+    WallTime lastModificationTime;
+    if (m_unifiedOriginStorageLevel == UnifiedOriginStorageLevel::None) {
```
switch

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    if (m_totalUsage)
+        m_totalUsage = *m_totalUsage + spaceRequested;
+
+    if (!m_totalUsage || *m_totalUsage > *m_totalQuota)
```
Why scheduling eviction if m_totalUsage is std::nullopt?

szewai replied:
To initialize `m_totalUsage` in `donePrepareForEviction`.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
-            return false;
-        });
+
+        for (auto clientOrigin : record.clientOrigins) {
```
auto&

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    uint64_t deletedOriginCount = 0;
+    while (*m_totalUsage > *m_totalQuota) {
+        ASSERT(!sortedOrigins.isEmpty());
+        auto topOrigin = sortedOrigins.takeLast();
```
What guarantees sortedOrigin is not empty?
I would tend to iterate on sortedOrigins and exit early if total usage is less than quota.

szewai replied:
Because m_totalUsage is calculated by summing up usage of sortedOrigins. Sure.

youennf commented on Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
```diff
@@ -50,11 +50,11 @@ class OriginQuotaManager : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPt
     bool grantWithCurrentQuota(uint64_t spaceRequested);
     bool grantFastPath(uint64_t spaceRequested);
-    void usageUpdated(uint64_t newUsage);
+    void spaceGranted(uint64_t spaceRequested);
```
The function definition uses amount and I tend to prefer amount instead of spaceRequested.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,168 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+
+    std::sort(sortedOrigins.begin(), sortedOrigins.end(), [&](const auto& a, const auto& b) {
+        return originRecords.get(a).lastAccessTime > originRecords.get(b).lastAccessTime;
+    });
```
I would tend to have performEviction take a Vector<std::pair< SecurityOriginData , AccessRecord>> so that we do not need both originRecords and sortedOrigins.
We would then be able to sort more efficiently, since we would not query the map every time we do a comparison.

szewai replied:
This means will need to convert `HashMap<WebCore::SecurityOriginData, AccessRecord>` to `Vector<std::pair< SecurityOriginData , AccessRecord>>` in `donePrepareForEviction`, but `performEviction` does not necessarily perform eviction -- it compares m_usage and m_quota again, so the conversion might not be needed.

I could convert `HashMap<WebCore::SecurityOriginData, AccessRecord>` to `Vector<std::pair< SecurityOriginData , AccessRecord>>` in `performEviction`, after comparing m_usage and m_quota. Do you think it's more efficient?

youennf replied:
Yes, the thing I am trying to avoid is query the map during sorting.
We should just need to query the item itself

youennf commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.cpp:
```diff
@@ -2024,6 +2024,20 @@ Vector<RegistrableDomain> ResourceLoadStatisticsDatabaseStore::allDomains() cons
+    ASSERT(!RunLoop::isMain());
+
+    HashMap<RegistrableDomain, WallTime> result;
+    auto domainDatas = domains();
```
temp var is not really needed.

youennf commented on Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.cpp:
```diff
@@ -2024,6 +2024,20 @@ Vector<RegistrableDomain> ResourceLoadStatisticsDatabaseStore::allDomains() cons
+    auto domainDatas = domains();
+    for (auto& domainData : domainDatas) {
+        auto lastAccessedTime = std::max(domainData.mostRecentUserInteractionTime, domainData.mostRecentWebPushInteractionTime);
+        result.add(domainData.registrableDomain, lastAccessedTime);
```
WTFMove(domainData.registrableDomain) maybe.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,166 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+    });
+}
+
+WallTime NetworkStorageManager::lastModificationTimeForOrigin(const WebCore::ClientOrigin& origin, OriginStorageManager& manager)
```
manager can be const maybe?

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
```diff
@@ -339,55 +339,166 @@ void NetworkStorageManager::writeOriginToFileIfNecessary(const WebCore::ClientOr
+
+    Vector<WebCore::SecurityOriginData> sortedOrigins;
+    for (auto& origin : originRecords.keys())
+        sortedOrigins.append(origin);
```
auto sortedOrigins = copyToVector(originRecords, [] (auto& keyVakue) ... );

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -225,8 +222,22 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver {
+    void schedulePerformEviction();
+    struct AccessRecord {
+        bool isActive { false };
+        uint64_t usage;
```
Let's also set this one to 0 for good measure.

youennf commented on Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
```diff
@@ -247,7 +258,9 @@ class NetworkStorageManager final : public IPC::WorkQueueMessageReceiver {
     std::optional<uint64_t> m_volumeCapacityOverride;
-    std::unique_ptr<NetworkQuotaManager> m_quotaManager;
+    std::optional<uint64_t> m_totalUsage;
+    std::optional<uint64_t> m_totalQuota;
```
I guess we do not really need m_totalQuota to be an optional. It could be a uint64_t and 0 means uninitialized.

szewai committed:
Make data eviction based on user interaction time when ITP is enabled
https://bugs.webkit.org/show_bug.cgi?id=255272

Reviewed by Youenn Fablet.

Revamp data eviction from exceeding total quota. The major changes:
1. NetworkStorageManager tracks total usage and performs eviction by itself instead of relying on NetworkQuotaManager;
this is because NetworkStorageManager has a list of origins that are being used (m_originStorageManagers) and it's more
convenient to do operation on OriginStorageManager from NetworkStorageManager.
2. NetworkStorageManager uses how much space is granted to all origins as estimated total usage. It only calculates the
real usage when estimated usage is bigger than quota, and then uses real usage to decide if eviction is needed.
3. When ITP is enabled, NetworkStorageManager decides the eviction order of origins based on interaction time from ITP;
when ITP is not enabled, NetworkStorageManager decides the order based on last modification time of storage files or
directories.
4. NetworkStorageManager will update last modification time of storage files or directories of an origin when it is
granted space.

API test: WKWebsiteDataStoreConfiguration.TotalQuotaRatioWithResourceLoadStatisticsEnabled

* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.cpp:
(WebKit::ResourceLoadStatisticsDatabaseStore::allDomainsWithLastAccessedTime const):
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsDatabaseStore.h:
* Source/WebKit/NetworkProcess/Classifier/ResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.cpp:
(WebKit::WebResourceLoadStatisticsStore::registrableDomainsWithLastAccessedTime):
* Source/WebKit/NetworkProcess/Classifier/WebResourceLoadStatisticsStore.h:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::registrableDomainsWithLastAccessedTime):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::createNetworkStorageManager):
(WebKit::NetworkSession::NetworkSession):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::create):
(WebKit::NetworkStorageManager::NetworkStorageManager):
(WebKit::NetworkStorageManager::writeOriginToFileIfNecessary):
(WebKit::NetworkStorageManager::spaceGrantedForOrigin):
(WebKit::NetworkStorageManager::schedulePerformEviction):
(WebKit::NetworkStorageManager::prepareForEviction):
(WebKit::NetworkStorageManager::lastModificationTimeForOrigin const):
(WebKit::NetworkStorageManager::donePrepareForEviction):
(WebKit::NetworkStorageManager::performEviction):
(WebKit::NetworkStorageManager::originStorageManager):
(WebKit::NetworkStorageManager::updateLastModificationTimeForOrigin):
(WebKit::NetworkStorageManager::quotaManager): Deleted.
(WebKit::NetworkStorageManager::updateOriginModificationTime): Deleted.
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.cpp:
(WebKit::OriginQuotaManager::create):
(WebKit::OriginQuotaManager::OriginQuotaManager):
(WebKit::OriginQuotaManager::usage):
(WebKit::OriginQuotaManager::grantWithCurrentQuota):
(WebKit::OriginQuotaManager::spaceGranted):
(WebKit::OriginQuotaManager::grantFastPath):
(WebKit::OriginQuotaManager::usageUpdated): Deleted.
* Source/WebKit/NetworkProcess/storage/OriginQuotaManager.h:
(WebKit::OriginQuotaManager::create):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.cpp:
(WebKit::OriginStorageManager::createQuotaManager):
(WebKit::OriginStorageManager::OriginStorageManager):
* Source/WebKit/NetworkProcess/storage/OriginStorageManager.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebsiteDatastore.mm:
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/262900@main
